### PR TITLE
Replace std::abort() with std::runtime_error

### PR DESCRIPTION
**Pull request by norihiro-w**
as issued in #1165. I added ERRMSG macro (see 74bd882ff513eeb8ebdbe68762e87e2bbb34a83a) in StringTools.h so that one can write as

```
throw std::runtime_error(ERRMSG("Some error with type %d", -1));
```

which generates an error message "Some error with type -1 at (file name), line (line no)".

### CODE REVIEW

**Comment by chleh** [comment]
Wouldn't it be better to introduce a macro `OGS_ERROR(...)` or `OGS_FATAL(...)` instead of always having to write `throw std::runtime_error(ERRMSG(...))`?

File: BaseLib/StringTools.cpp [hunk 1]
```diff
@@ -141,3 +142,20 @@ void correctScientificNotation(std::string filename, std::size_t precision)
...
+{
+    char buffer[256];
+    int size_of_buffer = sizeof(buffer) / sizeof(char);
+    memset(buffer, 0, sizeof(buffer));
+    va_list args;
+    va_start(args, format_str);
+#ifdef MSVC
+    vsnprintf_s(buffer, size_of_buffer, _TRUNCATE, format_str.data(), args);
+#else
+    vsnprintf(buffer, size_of_buffer, format_str, args );
```
Comment (chleh) [inline BaseLib/StringTools.cpp]: Error messages of more than 256 characters are pretty long, but I think there should be a check if the buffer was long enough, and if not a bigger buffer should be allocated dynamically (cf. "Notes" in http://en.cppreference.com/w/cpp/io/c/fprintf).

File: BaseLib/StringTools.cpp [hunk 2]
```diff
@@ -141,3 +142,20 @@ void correctScientificNotation(std::string filename, std::size_t precision)
...
+
+
+std::string format(const char* format_str, ... )
+{
+    char buffer[256];
+    int size_of_buffer = sizeof(buffer) / sizeof(char);
+    memset(buffer, 0, sizeof(buffer));
+    va_list args;
+    va_start(args, format_str);
+#ifdef MSVC
```
Comment (chleh) [inline BaseLib/StringTools.cpp]: Only wondering: Is this necessary?

File: BaseLib/ConfigTree.cpp [hunk 3]
```diff
@@ -7,9 +7,12 @@
  *
  */
 
-#include <logog/include/logog.hpp>
 #include "ConfigTree.h"
 
+#include <logog/include/logog.hpp>
+
+#include "StringTools.h"
```
Comment (chleh) [inline BaseLib/ConfigTree.cpp]: I would like a separate `BaseLib/Logger.h` header, which includes logog and StringTools, better.

**Comment by chleh** [comment]
Will the error message of the exception always be printed? (I.e., in debug and release builds?)
Or do we need a dedicated try/catch block in the main function for that?

**Comment by chleh** [comment]
After the questions are answered :+1:

**Comment by norihiro-w** [comment]
- Having `OGS_ERROR` macro is also fine, though explicitly writing std::runtime_error makes it clear what is really happening. I would like to hear how others think about it.
- I will add a try/catch block in the main function. IMO those error messages should be printed also in release mode.

File: BaseLib/StringTools.cpp (hunk 1 above)
Comment (norihiro-w) [inline BaseLib/StringTools.cpp]: will check it

File: BaseLib/StringTools.cpp (hunk 2 above)
Comment (norihiro-w) [inline BaseLib/StringTools.cpp]: will check it.TCLAP has also this condition

File: BaseLib/ConfigTree.cpp (hunk 3 above)
Comment (norihiro-w) [inline BaseLib/ConfigTree.cpp]: i am not sure if this is a good idea. exceptions are different from logging.

**Comment by endJunction** [comment]
At first I agreed with Christoph having a single macro like OGS_FATAL. This could be also replaced with something different than throw. On the other hand there are more exception types than the runtime error; having a single macro takes the opportunity to have different error handling...

**Comment by chleh** [comment]
But catching exceptions occasionally is discouraged, AFAIK...

**Comment by norihiro-w** [comment]
how about a macro `OGS_RUNTIME_ERROR`? In the future, someone could add `OGS_LOGICAL_ERROR` or something else.

**Comment by chleh** [comment]
OGS_RUNTIME_ERROR is OK.
But is there the chance that anyone of us will ever make a distinction between different kinds of exceptions thrown?

**Comment by norihiro-w** [comment]
no idea yet

**Comment by norihiro-w** [comment]
or shall we use `OGS_FATAL` now and rename it later if someone wants to use different exceptions?

**Comment by norihiro-w** [comment]
updated. please check each commit

**Comment by chleh** [comment]
perfect. :+1:

**Pull request merged**